### PR TITLE
jinja templates in configs

### PR DESCRIPTION
**Pull request by jpeletier**
# What does this implement/fix?

This is a functioning proposal to have templates in ESPHome configuration files in order to simplify complex configurations and allow for code reusability. I chose Jinja since it is well-known in the HomeAssistant community and other YAML-based configuration languages such as Ansible.

This change is opt-in and backwards compatible. Existing configurations are not affected.

The PR will consider any string in the configuration a potential Jinja template. The substitution syntax (`${ }`) is expanded to allow also for jinja expressions. Thus, variable names or expressions inside `${ }` will be evaluated and the result of the expression replaced into the string.

Jinja can perform math and render number values (for example to calculate modbus register addresses) or even **lists or full pieces of configuration** based on substitution variables in the top-level configuration file, command-line substitutions as well as `!include` variables (as per https://github.com/esphome/esphome/pull/3510), since these values are available to Jinja as Jinja variables.

This implementation covers all the cases described in https://github.com/esphome/feature-requests/issues/805 and builds on the current packages, substitutions and `!include` implementation, while maintaining backwards compatibility.

### `!for`

`!for` will iterate over a range and generate a list of objects based on the `repeat` template given.
#### Example:

```yaml
binary_sensor: !for
    items: [33, 36, 39]
    var: pin
    repeat:
      platform: gpio
      pin: ${pin}
      name: GPIO-${pin}
      device_class: door
```
The above is equivalent to writing the following:
```yaml
binary_sensor:
  - platform: gpio
    pin: 33
    name: GPIO-33
    device_class: door
  - platform: gpio
    pin: 36
    name: GPIO-36
    device_class: door
  - platform: gpio
    pin: 39
    name: GPIO-39
    device_class: door
```
Furthermore, the `items` parameter can be an expression too, or contain a substitution variable:
```yaml
substitutions:
  detector_pins: [33, 36, 39]

binary_sensor: !for
    items: "${ detector_pins + [14, 15] }"
    var: pin
    repeat:
      platform: gpio
      pin: ${pin}
      name: GPIO-${pin}
```
Also, you are not constrained to simple objects in the range:
```yaml
substitutions:
  zones:
    - name: Living Room
      code: 5
    - name: Kitchen
      code: 7

sensor: !for
    items: zones
    var: zone
    repeat:
      platform: modbus_controller
      address: "${ (zone.code - 1) * 4 }"
      name: "${ zone.name } Temperature"
      device_class: temperature
```
### `!if`
`!if` will evaluate a condition and if true will render the object under `then`, otherwise it will render the object under `else`:
#### Example:
```yaml
substitutions:
  enable_special_sensor: true

binary_sensor:
  - platform: gpio
    pin: 14
    name: always_enabled_sensor
  - !if
    condition: "${ enable_special_sensor }"
    then:
      platform: gpio
      pin: 15
      name: "special sensor"
    else:
      platform: gpio
      pin: 13
      name: "regular sensor"
```
As before, `condition` can contain variable names or any expression. Any nonzero value counts as `true`. The value `0`, empty strings, empty lists or empty objects count as `false`
`else` can be omitted. In that case, no object is rendered if the condition is false

### helper tag: `!literal`
`!literal` tells the parser to not interpret the text as Jinja, thus ignoring any `{{ }}` values. Note that substitutions will still be executed.
#### Example:
```yaml
binary_sensor:
  -   platform: gpio
      pin: 15
      name: !literal "Sensor with ${cool} name"
```

## Example configuration
Consider a manufacturer of an AC unit "Kool, Inc". Kool provides a modbus interface to manage the fancoil units and physical thermostats in each room of the house. Depending on how the physical structure of the house is, the company will install the units and assign each of the rooms a numeric code in the modbus interface. In the modbus documentation, the registers to manage different aspects of each room are a function of that numeric code. The different functions are current room temperature, target temperature and on/off switch for each zone.

Kool, Inc decides to support ESPHome and provides a package in the form of a yaml file or GitHub package so as users we can focus on just defining the names of the rooms and not have to worry about anything else. This package requires 3 parameters: the modbus component id, the modbus slave address and house zones definition. Thus, the package is instantiated like this for our 4-room house:

`configuration.yaml`:
```yaml
esphome:
  name: "test"

esp32:
  board: esp32dev
  variant: esp32
  framework:
    type: arduino

packages:
  hvac: !include
    file: "kool_hvac_controller.yaml"
    vars:
      modbus_id: modbus1
      modbus_address: 0x49
      controller_id: kool_hvac
      zones:
        - name: Living Room
          code: 3
        - name: Kitchen
          code: 7
        - name: Bedroom
          code: 4
        - name: Office
          code: 5
# [...]
```
Nothing new above, except that when defining vars for the package we are not restricted to strings anymore. Arbitrarily complex objects can now be used, like the `zones` parameter above that describes the rooms of the house.

Now let's look at the vendor-provided package file: `kool_hvac_controller.yaml`:
```yaml
modbus_controller:
  - id: ${controller_id} # substitution-type syntax, still valid
    address: "${ modbus_address }" # new jinja syntax
    modbus_id: ${modbus_id}

# In this example we iterate over the `zones` parameter to render the configuration
# for target temperature adjustment
number: !for
    items: zones
    var: zone
    repeat:
    - platform: modbus_controller
        name: "${ zone.name } Target Temperature"
        id: target_temp{{ zone.code }}
        modbus_controller_id: "${ controller_id }"
        address: "${ (zone.code - 1)*4 + 1 }" #compute modbus address according to the hvac unit documentation
        value_type: U_WORD
        min_value: 15
        max_value: 35
        step: 0.5
        unit_of_measurement: °C
        lambda: return x*0.5;
        write_lambda: return x*2;

# In this other example we render temperature sensors by 
# invoking a temperature sensor
# definition that is in another file, using !include
# we also show that with a conditional statement we 
# can add the sensor or not.
sensor: !if
  condition: ${ enable_temp_sensor }
  then: !include
    file: "temp_sensor.yaml"
    vars:
      zone: "${ zone }"
      controller_id: "${ controller_id }"

# In this example we mix fixed elements with
# elements generated by `!for`
switch:
    - platform: modbus_controller
      name: Master Activation
      modbus_controller_id: ${ controller_id }
      register_type: holding
      address: 1000
      bitmask: 1
    - !for
      items: zones
      var: zone
      repeat:
        - platform: modbus_controller
          name: "Activate Cooling ${ zone.name }"
          id: "cooling_{{ zone.code }}"
          modbus_controller_id: "${ controller_id }"
          register_type: holding
          address: "${ (zone.code - 1)*4 }"
          bitmask: 1

```

`temp_sensor.yaml`
```yaml
platform: modbus_controller
name: ${zone.name} Current Temperature
address: ${ (zone.code - 1) * 4 + 2 }
register_type: holding
id: "current_temp${ zone.code }"
modbus_controller_id: ${controller_id}
state_class: measurement
device_class: temperature
unit_of_measurement: °C
accuracy_decimals: 1
```

## Trying this
You can check out this PR and test jinja right now!. I would like to get some feedback about it to see if this is the right way to implementing templates and how it would simplify your configurations.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

https://github.com/esphome/feature-requests/issues/805

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by ssieb** [comment]
Will this conflict with https://esphome.io/components/api.html#homeassistant-service-action ?

**Comment by jpeletier** [comment]
@ssieb yes, it would conflict. It is an important discussion, because we have different options I think:

## Option 1. Do nothing (breaking change)

Users of `homeassistant.service` action would have to opt out of ESPHome jinja by tagging their strings with `!string`, which disables string expansion:
```yaml
substitutions:
  controller_name: MyController
  controller_id: 76

# [....]

  - homeassistant.service:
      service: notify.html5
      data:
        title: "{{ controller_name }}-{{ controller_id}}" # processed by ESPHome's jinja
      data_template:
        message: !string "The humidity is {{ my_variable }}%." #ESPHome jinja disabled here
      variables:
        my_variable: |-
          return id(my_sensor).state;
```

## Option 2. Make templating opt-in (jinja is disabled by default)

Everything works as it was (jinja is disabled), except for keys and subkeys of `!if` and `!for` and a new one we'd call `!jinja` which enables it for children of the tagged key. (You still can disable locally with `!string`)

```yaml
substitutions:
  controller_name: MyController
  controller_id: 76

# [....]

  - homeassistant.service:
      service: notify.html5
      data:
        title: !jinja "{{ controller_name }}-{{ controller_id}}" # only this item would use ESPHome's jinja
      data_template:
        message: The humidity is {{ my_variable }}%.
      variables:
        my_variable: |-
          return id(my_sensor).state;
```
or (disabling locally):

```yaml
  - homeassistant.service: !jinja #enable Jinja for this and all subobjects
      service: notify.html5
      data:
        title: "{{ controller_name }}-{{ controller_id}}" # would use ESPHome's jinja
      data_template:
        message: !string "The humidity is {{ my_variable }}%." # disabled locally
      variables:
        my_variable: |-
          return id(my_sensor).state;
```

## Option 3. Like option 2, but a global configuration key enables or disables jinja (disabled by default)

```yaml
esphome:
  enable_jinja: true

[...]

  - homeassistant.service:
      service: notify.html5
      data:
        title: "{{ controller_name }}-{{ controller_id}}" # would use ESPHome's jinja
      data_template:
        message: !string "The humidity is {{ my_variable }}%." # disabled locally
      variables:
        my_variable: |-
          return id(my_sensor).state;
```

**Comment by OttoWinter** [comment]
I like the use of YAML tags for the `!if` and `!for` statements in this PR! Thank you @jpeletier for driving this effort forward, I really like the approach here!

I'll post some of my initial thoughts here:

---

For substitution syntax: One idea would be to just deprecate/remove the substitution syntax. My concern is that by having these two syntaxes coexist (when they essentially do the same thing), we end up creating more confusion long term (there should only be one preferred way to do something). But I'm unsure, since it would be a major breaking change and we'd probably need to write something that auto-suggests how to update the config.

---

Note: in jinja you can also set different start/end strings to enter templating; maybe one approach could be to use `${` and `}` to start/end jinja2 templates (see `block_start_string`, etc in jinja2.Environment); So like `address: "${ (zone.code - 1)*4 + 1 }"`. The advantage is that then we reuse the substitution syntax, but downside of course is that it may be confusing as users think of `{{ }}` for jinja2 syntax.

---

Yeah `homeassistant.service` will be a challenge. However, it's probably the only place these `{{` show up in normal configs, and I think we can fix 90% of frustration by just having a good error message pointing to the ability to use `!string` and good docs.

Ansible of course uses this opt-out approach to templating too, and I like it a lot there. Initially, I'd vote for option 1.

---

Re `!string`: I think this name is confusing, because when I type it without `!string`, isn't that also a string then (I know what it's supposed to mean of course, but as a user it could be confusing). Perhaps a better name could be `!raw`?

---

Re `!for` keys: One disavantage of `!for` is that we can't use jinja2 syntax like `{% for x in zones %} ... {% endfor %}`, but that could be quite nice because then you get the jinja2 filters and all.

Probably not possible, but do you think a syntax like this could be made to work somehow?

```yaml
number: !for
    items: `{% for zone in zones %}
    repeat:
    - platform: modbus_controller
      asdf: "{{ zone }}"
```

---

If you haven't looked at ansible already, I'd recommend looking a bit through their [loop syntax](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html) which is quite similar (sans the YAML tags). For example, I think their approach of using `item` as the default `!for` loop var name could be useful for us too.

**Comment by jpeletier** [comment]
> I like the use of YAML tags for the `!if` and `!for` statements in this PR! Thank you @jpeletier for driving this effort forward, I really like the approach here!

Thanks!
 
> For substitution syntax: One idea would be to just deprecate/remove the substitution syntax. My concern is that by having these two syntaxes coexist (when they essentially do the same thing), we end up creating more confusion long term (there should only be one preferred way to do something). But I'm unsure, since it would be a major breaking change and we'd probably need to write something that auto-suggests how to update the config.
> Note: in jinja you can also set different start/end strings to enter templating; maybe one approach could be to use `${` and `}` to start/end jinja2 templates (see `block_start_string`, etc in jinja2.Environment); So like `address: "${ (zone.code - 1)*4 + 1 }"`. The advantage is that then we reuse the substitution syntax, but downside of course is that it may be confusing as users think of `{{ }}` for jinja2 syntax.

The approach you mention (keep to `${ expression }`) has a number of important advantages I think:
- Does not conflict with `homeassistant.service` and allows to mix.
- Reuse substitution syntax. Backward compatible. People evolve from simple substitutions to jinja expressions without noticing. I think this is a strong point.
- Save a lot of quotes, namely:

```yaml
- platform: modbus
  name: ${name}
```
is valid, whereas:
```yaml
- platform: modbus
  name: {{name}}
```
Doesn't work. You have to put quotes `name: "{{name}}"`, otherwise the YAML parser chokes on `{` as the first character in a string value. So `{{ }}` is actually a disadvantage.

Disadvantages:
- Looks less "jinjish", however, why would Jinja allow such customization then?

> Ansible of course uses this opt-out approach to templating too, and I like it a lot there. Initially, I'd vote for option 1.

I am also inclined for opt 1 initially but maybe keeping `${ expression }` does the job and we get the best of both worlds.
 
> Re `!string`: I think this name is confusing, because when I type it without `!string`, isn't that also a string then (I know what it's supposed to mean of course, but as a user it could be confusing). Perhaps a better name could be `!raw`?

I'd save the `!raw` word for the future, too generic. What  about `!literal` ?

> Re `!for` keys: One disavantage of `!for` is that we can't use jinja2 syntax like `{% for x in zones %} ... {% endfor %}`,

Actually something like this **already works**. Since any string is considered a template, you can ignore `!for` and do the following:

```yaml
number: |
  {% set sensors = [] %}
  {% for zone in zones %}
  {% do sensors.append({
    "platform": "modbus_controller",
    "address": ( zone.code - 1) * 4,
    "modbus_controller_id": controller_id,
    "name": zone.name ~ " Temperature"}) %}
  {% endfor %}
  {{ sensors }}
```
is equivalent to:
```yaml
number: !for
  range: zones
  var: zone
  repeat:
    - platform: modbus_controller,
      address: "{{ (zone.code -1) * 4}}"
      modbus_controller_id: "{{ controller_id }}",
      name: "{{ zone.name }} Temperature"
```

For a more YAML-looking interaction, I'd have to re-introduce `!eval`, which I removed after initial feedback. See full examples in [the old PR](https://github.com/esphome/esphome/pull/3602). In essence:

```yaml
number: !eval |
  {% for zone in zones %}
  - platform: modbus_controller
    address: {{ (zone.id-1)*4 + 1 }}
    name: {{ zone.name }} Target Temperature
    modbus_controller_id: {{ controller_id }}
  {% endfor %}
```
This is equivalent to above as well. What `!eval` does is **runs the YAML parser over the returned string**. The disadvantage is that an indentation mistake can make things complicated, since the IDE doesn't know that string is YAML, as pointed by @quentinmit during a discussion in Discord.

> but that could be quite nice because then you get the jinja2 filters and all.

Filters already work within `{{}}`, e.g.
```yaml
address: "{{ '12.5' | int(default=0) }}"
```
stores the `Int` value `12` in `address` and is equivalent to:
```yaml
address: 12
```
Note that, like Ansible, the return *type* of a template is the return type of the expression, if there is no text before the fist `{`.

> Probably not possible, but do you think a syntax like this could be made to work somehow?
> 
> ```yaml
> number: !for
>     items: `{% for zone in zones %}
>     repeat:
>     - platform: modbus_controller
>       asdf: "{{ zone }}"
> ```

Not sure what you mean since you didn't close the loop. But `!for` does not remove jinja capability. Therefore the `items` key (I called it `range`) accepts a list, and that list can come from anywhere. It can be a literal with no jinja:
```yaml
range:
  - name: "zone1"
    code: 1
  - name: "zone2"
    code: 2
  - name: "zone1"
    code: 3
```
a variable in the current scope (substitution or package variable):
```yaml
range: zones
```
or a `{{}}` expression with or without filters:
```yaml  
range: "{{ [{'name':'zone1', 'code': 1},{'name':'zone2', 'code': 2},{'name':'zone3', 'code': 3}] | sort(attribute='code') }}"
```
or a full Jinja template with blocks:

```yaml
substitutions:
  codes: [1,2,3]

number: !for
    range: |
      {% set zones = [] %}
      {% for code in codes %}
      {% do zones.append({"name": "zone" ~ code, "address": (code-1)*4}) %}
      {% endfor %}
      {{ zones }}
    repeat:
      - platform: modbus_controller
        address: "{{ item.address }}"
        name: "{{ item.name }}"

```
The above is a bit of a contrived example, but first `range` has to be calculated. For that, Jinja executes the template and produces the list, which is then processed by `!for` to iterate over...
Note that this is not only for `range`. Any key in the configuration can have any type of jinja, so long as the provided jinja returns a type accepted by the component validators later.

> If you haven't looked at ansible already, I'd recommend looking a bit through their [loop syntax](https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html) which is quite similar (sans the YAML tags). For example, 

Yes, I went through it, but found their syntax confusing, though I extracted what I thought was easy to understand for people in general.

> I think their approach of using `item` as the default `!for` loop var name could be useful for us too.

I copied that!. If you omit `var`, then the default variable name is `item`.

## In summary:

I would:
* Go with backward-compatible `${ expression }` syntax instead of `{{ expression }}`
* Rename `!string` to `!literal`
* Keep `!if` and `!for` since they simplify and make things readable and YAML-looking
* Do not re-introduce `!eval` for now.

**Comment by OttoWinter** [comment]
> Doesn't work. You have to put quotes name: "{{name}}", otherwise the YAML parser chokes on { as the first character in a string value. So {{ }} is actually a disadvantage.

Ah yes that's an added advantage. I think this is also why [GitHub Actions does it with `${{ }}`](https://docs.github.com/en/actions/learn-github-actions/expressions).

Just a wild idea, but maybe `${}` AND `${{ }}` is an option (where the former only allows vars, and latter is for full jinja2 expressions). It would feel more "jinjish" again, and might be easier to understand/convey than the `${}` & `{{ }}` combo.

> Looks less "jinjish", however, why would Jinja allow such customization then?

I first stumbled on those options some time ago when I built [templating for some LaTeX code](http://eosrei.net/articles/2015/11/latex-templates-python-and-jinja2-generate-pdfs) (where `\BLOCK{` feels more natural). Or for languages where `{{` occurs often by itself, and so would hinder jinja2.

Anyway, even if jinja allows those options, except for LaTeX I've never seen it with syntax other than the `{{ }}`. And I fear users (who, let's face it, don't RTFM), won't discover they can use jinja2 syntax here. I'd think this is probably some complex bash-like syntax. Maybe `${{ }}` strikes a balance there?

> Actually something like this already works. Since any string is considered a template, you can ignore !for and do the following:

Ah, thanks, didn't think of that!

> What !eval does is runs the YAML parser over the returned string. The disadvantage is that an indentation mistake can make things complicated, since the IDE doesn't know that string is YAML, as pointed by 

Ah yes we had a couple of those PRs over the years, but yeah the spacing issue always made this a show-stopper.

> I'd save the !raw word for the future, too generic. What about !literal ?

Sounds good

**Comment by jpeletier** [comment]
> Just a wild idea, but maybe `${}` AND `${{ }}`  (where the former only allows vars, and latter is for full jinja2 expressions)

Why would we restrict `${}` to only vars? Users that think `${}` only works with vars will be happy, given that `${ var }` in Jinja (once configured) coincidentally produces the same result as a substitution. This is very convenient and backwards compatible. We should enjoy our lucky day here! 😉

Having two options would make implementation and maintenance more complicated, because to implement `${ expression }` is just changing Jinja config and that's it, no more code, and then we can do away with the current substitution code, because Jinja will do the substitutions itself.

And if users don't read the manual and don't find out they can put jinja in there is because they didn't really need it. I think that users that need this will end up discovering it. And I will make sure they do in the documentation.

So if you agree, I would do the following:

- Use `${}` for now. We can always introduce `${{ }}` and/or `{{ }}` later if really needed, even as an alias.
- Use `!literal`
- Keep `!for` and `!if`
- Document all features with detail

**Comment by jpeletier** [comment]
@OttoWinter This should be ready for review. Here are the changes:

* Universal syntax `${ expression }` does jinja and substitutions and is backward-compatible.
* Substitutions may contain values other than strings, including numbers, lists or otherwise any complex object.
* Removed old substitutions code and virtual component. Substitutions are now all executed during YAML parsing as part of Jinja. This comprises `!include` and `!for` loop variables, taking into account the scopes of those variables, i.e., the iterator variable of `!for` is only valid in children keys of that `!for`, and shadows any existing variable with that name within its scope.
* Command-line substitutions can now contain types other than strings, that is, you can write `esphome -s pins '[14,15,33,36]' compile example.yaml` and the `pins` substitution variable will contain a list.
* Renamed `!for` `range` to `items`, e.g.:
```yaml
substitutions:
  pins: [14,15,0,2]
binary_sensor: !for
  items: ${pins + [33,36]}
  var: pin
  repeat:
    platform: gpio
    pin: ${pin}
```
* Bonus: Added `vars` for remote packages. This allows specifying parameters for packages in remote repositories:
```yaml
substitutions:
  sensor_name: MyName

packages:
  test:
    url: https://github.com/jpeletier/testpackage.git
    ref: master
    file: test.yaml
    vars: # new: this way you can pass values to the remote package
      sensor_name: ${sensor_name} Super Sensor
      pin: 4
```
`vars` under a remote package override vars defined in `substitutions`, but only for that package.
* Added tests for `!for`, `!if`, `!literal` and jinja
* `yaml_util.py` was growing too large, so I moved the `make_data_base` code to a different file.

I hope this is a good addition to simplify configurations and make them more reusable!. Please let me know of any comments

**Comment by bearpawmaxim** [comment]
Hi @jpeletier! Is it possible to achieve overriding of items from packages with this? Like in #3555

**Comment by jpeletier** [comment]
@bearpawmaxim This PR does not provide item overriding, but it is true that you could achieve similar functionality. 

Let's look at some examples coming from [#3555](https://github.com/esphome/esphome/pull/3555) and how would them work with this PR:
```yaml
packages:
  parent:
    sensor:
    - id: uptime_sensor
      platform: uptime
sensor:
- id: !extend uptime_sensor
  update_interval: 10s
```
Would become:
```yaml
# main.yaml
packages:
  parent: !include
    file: uptimesensor.yaml
    vars:
      interval: 10s
# -------
# uptimesensor.yaml
sensor:
  - id: uptime_sensor
    platform: uptime
    update_interval: ${interval}
```
Actually, the above does not use anything from this PR, but from the already released [#3510](https://github.com/esphome/esphome/pull/3510) (vars in `!include`).

But let's look at [this other example](https://github.com/esphome/esphome/pull/3555#discussion_r901821977) from [#3555](https://github.com/esphome/esphome/pull/3555):
```yaml
packages:
  base: !include base.yaml
  rev2: !include rev2.yaml  # some boards are rev 2 hardware

# --------
# base.yaml
sensor:
  - platform: dht
    id: asdf
    pin: GPIO2
# --------
# rev2.yaml
sensor:
  - platform: dht
    id: !extend asdf
    pin: GPIO23
```
Would be rewritten with the new functionality in this PR as:
```yaml
packages:
  base: !include
    file: base.yaml
    vars:
      revision: 2
---
# base.yaml
sensor:
  - platform: dht
    id: asdf
    pin: !if
      condition: ${ revision < 2 }
      then: GPIO2
      else: GPIO23
```
Note how the package author hides away details about exact GPIOs and just provides a "revision" parameter for you to use externally.

My conclusion is you might not need `!extend` if the package is written well enough to accomodate your configuration needs.

But, what if the package author did not provide a way to configure and you just want to hack it? Here is the way. Let's go back to this simple example, in which the author did not provide a way to set `update_interval`:

```yaml
packages:
  parent: !include "uptimesensor.yaml"
sensor:
- id: !extend uptime_sensor
  update_interval: 10s

#----
# uptimesensor.yaml
sensor:
- id: uptime_sensor
    platform: uptime
```
Let's say you want to be able to add `update_interval` hacking into the original package. Here is the way:
```yaml
packages:
  parent: !include
    file: "override.yaml"
    vars:
      original: !include "uptimesensor.yaml"
      update_interval: 10s
#----
# uptimesensor.yaml (untouched)
sensor:
- id: uptime_sensor
    platform: uptime

#----
# override.yaml
|-
  {% do original['sensor'][0].update({'update_interval' : update_interval }) %}
  ${ original }
```

It's a hack, but it works. The way it works is you load up the whole original package into the `original` variable, then run some jinja in `override.yaml` to set `update_interval` to the provided value. Woah! 😜

In fact, you could generalize `override.yaml` to be an utility package that helps work like `!extend` somehow.

**Comment by bearpawmaxim** [comment]
> It's a hack, but it works.

Dunno how the jinja works, but maybe there's some directives support? I mean, if we write `!extend:...` and jinja will do that array access and update call for us..
We even can write `!extend id` and jinja would merge everything after this directive, cause id is globally unique

**Comment by quentinmit** [comment]
You do realize that you're just talking about embedding arbitrary code-in-config, at that point, right? Obviously anything is _possible_ once you have arbitrary code execution, but I think these are really serving different use-cases.

I don't think the examples you pulled from #3555 are really the best choice for evaluating this. They were written to clearly demonstrate what it does, but they're not a good practical use. A more likely usecase for `!extend` is to add additional automations to a component, or filters, or some other larger block of config, in a way that the original author of a package didn't necessarily envision. I think the ergonomics of that are significantly worse than your example of setting a single field.

**Comment by jpeletier** [comment]
@quentinmit my last solution is ugly AF, just hacking around trying to answer the question from @bearpawmaxim :-)

**Comment by jpeletier** [comment]
@OttoWinter Any further comments on this PR. We'd really like to see this merged. Thanks

**Comment by jimtng** [comment]
> A more likely use case for `!extend` is to add additional automations to a component, or filters, or some other larger block of config, in a way that the original author of a package didn't necessarily envision. I think the ergonomics of that are significantly worse than your example of setting a single field.

I agree with @quentinmit here. #3555 makes it simpler to override packages, whereas the jinja way is too complicated. 

I love the jinja idea and I would like it be merged, but I don't see it as a practical replacement of #3555

Here's my use case:
```yaml
packages:
  watering: !include common/devices/watering-timer.yaml

time:
  - id: !extend ntp_time
    on_time:
      # always water in the morning
      - hours: 9
        minutes: 00
        seconds: 0
        then:
          - script.execute: run_scheduled_watering

      # In spring and summer, water in the afternoon too
      - months: JAN-MAR,SEP-DEC
        hours: 15
        minutes: 0
        seconds: 0
        then:
          - script.execute: run_scheduled_watering

#----
# watering-timer.yaml

time:
  - platform: sntp
    id: ntp_time
    timezone: ${timezone}
    servers: ["${ntp_server}", 0.pool.ntp.org, 1.pool.ntp.org]


```

**Comment by jpeletier** [comment]
Hi @jimtng 
I reached the conclusion with @quentinmit that `!extend` would be more convenient to extend packages that you didn't write.

However, if you are the package author, this PR would be an option to consider. Notice how the main configuration is simplified and organized, since everything related to `watering` resides in the package invocation, as opposed to having to declare the package and then putting things under a repeated `time:` key, potentially mixed with other unrelated things:

```yaml
packages:
  watering:
    file: common/devices/watering-timer.yaml
    vars:
      timezone: CET
      additional_servers: ["my.ntp-server.com"]
      schedule:
        - hours: 9 # always water in the morning
        - months: JAN-MAR,SEP-DEC # In spring and summer, water in the afternoon too
          hours: 15
```
And here you have the rewritten `watering-timer.yaml`:
```yaml
time:
  - platform: sntp
    id: ntp_time
    timezone: ${timezone or "GMT"}
    servers: ${ (additional_servers or []) + ["0.pool.ntp.org", "1.pool.ntp.org"] }
    on_time: !for
      items: schedule
      var: s
      repeat:
        - months: ${s.months or "*"}
          hours: ${s.hours or 0}
          minutes: ${s.minutes or 0}
          seconds: ${s.seconds or 0}
          then:
            - script.execute: run_scheduled_watering
```

Therefore, your main configuration is more about business-logic, whereas the package encapsulates the technical details.

**Comment by jpeletier** [comment]
@OttoWinter @jesserockz please consider this PR which would greatly simplify configs, there is a lot of work behind it. Thanks

**Comment by fxstein** [comment]
I want to add that with more complex scenarios and high numbers of devices, a solution like this quickly becomes necessary. I am not an expert on all the options discussed in this PR, but conditional logic is required to assemble complex configurations. I, too, have reached a point where my configs would greatly benefit from this functionality. Thank you all for contributing, and I hope this solution can get merged soon.

**Comment by jacauc** [comment]
Will this PR still be merged, or has it been abandoned?

**Comment by jpeletier** [comment]
@fxstein @jacauc I will fix conflicts and update this PR if I see approvals from @OttoWinter @jesserockz. This would indeed simplify complex configs, or remove the need for scripts that generate configs. Please vouch for it in Discord as well. Thanks

**Comment by jacauc** [comment]
> @fxstein @jacauc I will fix conflicts and update this PR if I see approvals from @OttoWinter @jesserockz. This would indeed simplify complex configs, or remove the need for scripts that generate configs. Please vouch for it in Discord as well. Thanks

New to discord, not sure how/where to vouch yet. Will poke around

**Comment by bojanpotocnik** [comment]
Any news on this (or https://github.com/esphome/feature-requests/issues/805)? As there are more and more devices in ESPHome, it makes sense to move as much config as possible to packages/includes... But this then requires either multiple very similar copies of the same package file or plenty of package files with minimal content, because there is no option to programmatically include/exclude some portion of the config in the package.

**Comment by jacauc** [comment]
> Any news on this (or [esphome/feature-requests#805](https://github.com/esphome/feature-requests/issues/805))? As there are more and more devices in ESPHome, it makes sense to move as much config as possible to packages/includes... But this then requires either multiple very similar copies of the same package file or plenty of package files with minimal content, because there is no option to programmatically include/exclude some portion of the config in the package.

I doubt this will happen

**Comment by bojanpotocnik** [comment]
I've just found [Packages as Templates](https://esphome.io/guides/configuration-types.html#packages-as-templates) (added with https://github.com/esphome/esphome-docs/pull/2980) which would make this even more useful.

For now, it seems like at least `!if` functionality can be partially mimicked through a makeshift use of `!extend` (https://github.com/esphome/esphome/pull/3555) and `!remove` (which is yet to be implemented - https://github.com/esphome/feature-requests/issues/2163).
